### PR TITLE
cbqn: 2022-11-27 -> 2023-02-01

### PR DESCRIPTION
**Pull request by skykanin**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by shnarazk — APPROVED** [review]
It is a shame that it doesn't support singeli.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 1]
```diff
@@ -13,52 +13,44 @@
 
 let
   # TODO: these submodules should be separated libraries
-  cbqn-bytecode-files = fetchFromGitHub {
-    name = "cbqn-bytecode-files";
+  cbqn-bytecode-submodule = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:16]: This is not a "submodule", is it?

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 2]
```diff
@@ -13,52 +13,44 @@
...
-    rev = "49c0d9a355698f54fff2c0caa177e2b341fabb45";
-    hash = "sha256-jm2ZzFxhr9o4nFR2rjYJz/4GH+WFnfU4QDovrOPI3jQ=";
+    rev = "05c1270344908e98c9f2d06b3671c3646f8634c3";
+    hash = "sha256-wKeyYWMgTZPr+Ienz3xnsXeD67vwdK4sXbQlW+GpQho=";
   };
 
-  nativeBuildInputs = [
-    pkg-config
-  ];
+  nativeBuildInputs = [ pkg-config ];
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:42]: Useless edit. Revert.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 3]
```diff
@@ -13,52 +13,44 @@
...
 
-  buildInputs = [
-    libffi
-  ];
+  buildInputs = [ libffi ];
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:44]: Useless edit. Revert.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 4]
```diff
@@ -13,52 +13,44 @@
...
 
   postPatch = ''
     sed -i '/SHELL =.*/ d' makefile
   '';
 
-  makeFlags = [
-    "CC=${stdenv.cc.targetPrefix}cc"
-  ]
-  ++ lib.optional enableReplxx "REPLXX=1";
+  makeFlags = [ "CC=${stdenv.cc.targetPrefix}cc" ]
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:52]: Useless edit. Revert.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 5]
```diff
@@ -74,43 +66,40 @@ stdenv.mkDerivation rec {
   '' + (if genBytecode then ''
     ${bqn-path} ./build/genRuntime ${mbqn-source} build/bytecodeLocal/
   '' else ''
-    cp ${cbqn-bytecode-files}/src/gen/{compiles,explain,formatter,runtime0,runtime1,src} build/bytecodeLocal/gen/
-  '')
-  + lib.optionalString enableReplxx ''
+    cp ${cbqn-bytecode-submodule}/gen/{compiles,explain,formatter,runtime0,runtime1,src} build/bytecodeLocal/gen/
+  '') + lib.optionalString enableReplxx ''
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:70]: Useless edit. Revert.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 6]
```diff
@@ -74,43 +66,40 @@ stdenv.mkDerivation rec {
...
-     install -Dm755 libcbqn${stdenv.hostPlatform.extensions.sharedLibrary} -t "$lib/lib"
-  ''
-  + ''
-     runHook postInstall
+    mkdir -p $out/bin/
+    cp BQN -t $out/bin/
+    # note guard condition for case-insensitive filesystems
+    [ -e $out/bin/bqn ] || ln -s $out/bin/BQN $out/bin/bqn
+    [ -e $out/bin/cbqn ] || ln -s $out/bin/BQN $out/bin/cbqn
+  '' + lib.optionalString enableLibcbqn ''
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:84]: Useless edit (except the indentation). Revert.

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 7]
```diff
@@ -74,43 +66,40 @@ stdenv.mkDerivation rec {
...
+  '' + ''
+    runHook postInstall
   '';
 
   meta = with lib; {
     homepage = "https://github.com/dzaima/CBQN/";
     description = "BQN implementation in C";
     license = licenses.gpl3Plus;
-    maintainers = with maintainers; [ AndersonTorres sternenseemann synthetica shnarazk ];
+    maintainers = with maintainers; [
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:95]: Why are you so inconsistent?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/interpreters/bqn/cbqn/default.nix (hunk 7 above)
Comment (skykanin) [reply pkgs/development/interpreters/bqn/cbqn/default.nix:95]: I used `nixpkgs-fmt` to do the formatting.

**Comment by skykanin** [comment]
> It is a shame that it doesn't support singeli.

I can try contributing this and refactoring all the submodules into separate derivations

File: pkgs/development/interpreters/bqn/cbqn/default.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/interpreters/bqn/cbqn/default.nix]: It is more akin to a subtree or branch. But whatever.

**Comment by skykanin** [comment]
@shnarazk ok singeli seems to be working now. Note that you need to override the `enableSingeli` flag because `stdenv.hostPlatform.avx2Support` is a pure evaluation which afaict always returns `false`.

File: pkgs/development/interpreters/bqn/cbqn/singeli.nix [hunk 8]
```diff
@@ -0,0 +1,37 @@
+{ lib
+, fetchFromGitHub
+, stdenvNoCC
+}:
+
+stdenvNoCC.mkDerivation {
+  pname = "Singeli";
+  version = "0.pre+date=2023-01-23";
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/singeli.nix:8]:
```suggestion
  pname = "singeli";
  version = "unstable-2023-01-23";
```

File: pkgs/development/interpreters/bqn/cbqn/singeli.nix [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+    mkdir -p $out/dev
+    cp -r $src $out/dev
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/mlochbaum/Singeli";
+    description = "High-level interface for low-level programming";
```
Comment (AndersonTorres) [inline pkgs/development/interpreters/bqn/cbqn/singeli.nix:32]:
```suggestion
    description = "A metaprogramming DSL for SIMD";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by shnarazk — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Many thanks for the hard work!

**Pull request merged**